### PR TITLE
Add support for SSH tunneling

### LINKED ISSUE #4: support ssh-tunneling
https://github.com/ember-cli/ember-cli-deploy/issues/108#issuecomment-81703926

### PR DESCRIPTION
**Pull request by tim-evans**
We set up ember-cli-deploy on our apps, then realized that SSH tunneling isn't supported. Instead of tossing the work, I ended up going in and adding support for tunneling to redis servers!

To do so, I've refactored the redis connection into a `redis-client` class that acts exactly like `then-redis` did. Deploy environments that need to SSH to deploy only need to add a `ssh` hash to their store. The `privateKey` is the only special change here- it refers to the path where the private key is rather than being the literal private key.

In addition to this work, I've upgraded the addon to be up-to-date with all of its dependent modules!

(This resolves #4)

### CODE REVIEW

**Comment by LevelbossMike** [comment]
Thx for this. I will have time to look into this on the weekend. Looks great!

**Comment by tim-evans** [comment]
@LevelbossMike Sweet!

FYI- I wasn't sure how to test tunneling. I have an idea now to setup a tunnel from localhost to localhost ;) If I have some time, I'll write that up.

File: lib/redis-client.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+function wrap(methodName) {
+  return function () {
+    var args = Array.prototype.slice.call(arguments);
+    return this._connection.then(function (client) {
+      return client[methodName].apply(client, args);
+    });
+  };
+}
+
+function promisifyNodeCallback(target, method) {
```
Comment (LevelbossMike) [inline lib/redis-client.js]: We are already using RSVP here anyway. There's [RSVP.denodify](http://emberjs.com/api/classes/RSVP.html#method_denodeify) to do the same thing.

**Comment by LevelbossMike** [comment]
again thx for this pr and sorry for taking so long to have a look at this. What's the reason again to move away from then-redis? Can't we use ssh-tunneling with then-redis directly?

File: lib/redis-client.js (hunk 1 above)
Comment (tim-evans) [inline lib/redis-client.js]: Ahh, yes. I'm not very familiar with node things :P

**Comment by tim-evans** [comment]
Yes, we could use then-redis directly.

I chose not to because:
- this solution provided minimal changes to the actual deployment code (just a single line change)
- the code for whether to SSH tunnel is isolated to a small portion of the redis client class

Mostly, I was having a bit of trouble deciphering how things worked :blush:, so I didn't want to change much in the actual deployment code.

**Comment by tim-evans** [comment]
Just rebased my changes onto master too

**Comment by LevelbossMike** [comment]
Thx!

**Comment by tim-evans** [comment]
:tada: Thank you! Any chance you could cut a new version as well?

**Comment by LevelbossMike** [comment]
:shipit: sure thing! I just released version 0.0.3. Thanks again!

**Comment by ghedamat** [comment]
:+1: this is amazing! thanks @tim-evans and @LevelbossMike I'm building a demo app right now for our Toronto meetup to show how easy this deploy strategy is, I'll test this right now :)

**Comment by gordonbisnor** [comment]
I'm not sure I understand the correct usage –– assuming I am deploying to a remote Linode server, if I don't include the SSH tunnel I am unable to connect. If I do include the SSH tunnel, I also need to add a passphrase option in my deploy.js file. I would rather not specify the password to my private key in plaintext. I'm sure that I am doing something wrong here –– would appreciate feedback! Or is there a way to pass that passphrase in when I run ember deploy, etc..

**Comment by LevelbossMike** [comment]
@gordonbisnor Because the deploy.js file is basically a js-script you can do whatever you want there. You can use https://www.npmjs.com/package/dotenv for managing secrets for example.

**Comment by gordonbisnor** [comment]
Ok thanks for the tip @LevelbossMike – I was looking at ember-cli-dotenv for this but wondered if there was anything already built-in!

**Comment by LevelbossMike** [comment]
You can use dotenv directly in deploy.js. ember-cli-dotenv is currently not supported (because there is a loading order issue between ember-cli-deploy and ember-cli-dotenv)

**Comment by gordonbisnor** [comment]
Okay, thanks @LevelbossMike that was where I was stumbling!

**Comment by thomassnielsen** [comment]
@gordonbisnor where did you specify the passphrase? I can't find any reference to it in this lib, or where it would pass it on to ssh2.

**Comment by gordonbisnor** [comment]
@thomassnielsen reading through various issues and/or code here, I learned that ember-deploy is farming out the process to another piece of software, I believe called tunnel-ssh, and from there I think stumbled upon the realization that you can add a passphrase, which will be passed along from the deploy.js file:

```
 store: {
      type: 'redis',
      ssh: {
        host: 'foo.com',
        username: 'foo',
        dstPort: 6379, // redis port
        dstHost: 'localhost', // redis host
        privateKey: '/path/to/key',
        passphrase: 'phrase-goes-here'
      }
    }
```

**Comment by thomassnielsen** [comment]
Thanks :smile:

**Pull request merged**